### PR TITLE
Rich printing

### LINKED ISSUE #92: Elisions and other rich printing
The unrepl printer provides elision: large locations are truncated and a *cough* hypermedia affordance *cough* is put where truncation happened; this hyperlink of sorts contains a regular form that you just have to evaluate to get back the remaining part... which itself can be elided if still too large. So it's related to #75.

Clients can render elisions as buttons and, on click, the button is replaced with more data. In practice it becomes a generic browsing facility (and Clojure 1.10 datafy/nav is going to make things even easier.) For example, unrepl printing of a random Java object contains a beans map of the object -- except this map is completely elided and will be created only when the user decides to navigate the object graph.

The catch is that an EDN representation of clojure values is returned. For example `#"abc"` is printed as `#unrepl/pattern "abc"` because regex patterns are not EDN. Same goes for rationals. The EDN representation allows to disambiguate between the printing of symbols and classes.

So it definitely adds some rendering complexity to the client, and cause one to reconsider how to do pretty printing when more data may appear in the middle of already pretty printed stuff (how to best reflow a pretty print?)

The best integration course seems tp make this an optional behaviour. You would opt for it by setting a custom `:printer`. 

The unrepl printer emits edn representations of clojure data

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by cgrand**
Fix #92.

Enable rich printing by setting `:printer` to `nrepl.elisions/printer` in an `:eval` message.

Long/infinite data, MIME attachments and `datafy`/`nav` are supported through elisions and custom literals.

Pretty printing has to be done on the client side.

### CODE REVIEW

File: src/clojure/nrepl/elisions.clj [hunk 1]
```diff
@@ -0,0 +1,44 @@
+(ns nrepl.elisions
```
Comment (bbatsov) [inline src/clojure/nrepl/elisions.clj]: I wonder if we shouldn't put this namespace and rich-printer under something like `nrepl.print` or whatever. Might be useful to structure the code like this if we ever come up with new printers.

Comment (bbatsov) [reply src/clojure/nrepl/elisions.clj]: Some short summary of the purpose of the ns and add `:added 0.6` metadata would be a nice addition as well.

File: src/clojure/nrepl/rich_printer.clj [hunk 2]
```diff
@@ -0,0 +1,460 @@
...
+  Features elisions, MIME attachments and general purpose object browsing.
+  Initially developed for unrepl."
+  (:refer-clojure :exclude [pr pr-str])
+  (:require
+   [clojure.core :as core]
+   [clojure.string :as str]
+   [clojure.edn :as edn]
+   [clojure.main :as main]))
+
+; even with limits on breadth and depth of printing, we can easily get big values:
```
Comment (bbatsov) [inline src/clojure/nrepl/rich_printer.clj]: I think this big comment can be converted to a docstring.

File: src/clojure/nrepl/rich_printer.clj [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+
+(defn ensure-defaults [bindings]
+  (let [bindings (merge-with #(or %1 %2) bindings defaults)]
+    (assoc bindings #'*print-budget*
+           (long (min (* 1N (bindings #'*print-level*) (bindings #'*print-length*)) Long/MAX_VALUE)))))
+
+(defprotocol MachinePrintable
+  (-print-on [x write rem-depth]))
+
+;; clojure 1.10 support
```
Comment (bbatsov) [inline src/clojure/nrepl/rich_printer.clj]: Or datafy/nav support? :-) Probably this will be a better description once 1.11 is out. :-)

**Comment by bbatsov** [comment]
Apart from the build failure my general feedback is:

* it'd be nice if the new printer had some unit tests, as I don't think one integration test is enough to cover well that it's working properly
* All public functions should ideally have some docstring. 
* We need to add a bit of documentation to the manual about tool writers and how they can leverage the new printer. 

Apart from this - the code looks very solid!

File: src/clojure/nrepl/elisions.clj (hunk 1 above)
Comment (bbatsov) [reply src/clojure/nrepl/elisions.clj]: Btw, why is this a separate ns to begin with? I wonder if it wouldn't better to have some `soft-store` ns and everything else to be in the `rich-print` ns.

File: src/clojure/nrepl/middleware/session.clj [hunk 4]
```diff
@@ -178,23 +178,21 @@
...
-            "clone" (register-session msg)
-            "close" (close-session msg)
-            "ls-sessions" (t/send transport (response-for msg :status :done
-                                                          :sessions (or (keys @sessions) [])))
-            (h msg)))))))
+    (if-some [the-session (if session
+                            (@sessions session)
+                            (create-session transport))]
+      (let [msg (-> msg (assoc :session the-session) (assoc-in [:print-options :nrepl/session] the-session))]
+        ;; TODO: yak, this is ugly; need to cleanly thread out-limit through to
```
Comment (bbatsov) [inline src/clojure/nrepl/middleware/session.clj]: Btw, do we still need this in light of the recent changes you've made?

File: src/clojure/nrepl/rich_printer.clj [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+        (when (and *print-meta* (meta x))
+          (write "]"))
+        (when (datafiable? x)
+          (write " ")
+          (set! *print-budget* (bump *print-budget* 1))
+          (print-on write (tagged-literal 'unrepl/... (*elide* (lazy-seq [(list (browsify x))]))) (inc rem-depth))
+          (write "]"))))))
+
+(defn base64-encode [^java.io.InputStream in]
+  (let [table "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/"
```
Comment (bbatsov) [inline src/clojure/nrepl/rich_printer.clj]: Shouldn't this be some constant?

File: src/clojure/nrepl/rich_printer.clj [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+        sb (StringBuilder.)]
+    (loop [shift 4 buf 0]
```
Comment (bbatsov) [inline src/clojure/nrepl/rich_printer.clj]: There a few magic numbers if this function - I wonder if we shouldn't give them some names. For me it's hard to figure out why those values are the ones needed.

File: src/clojure/nrepl/rich_printer.clj [hunk 7]
```diff
@@ -0,0 +1,460 @@
...
+                                                                                      (-> .getGraphics (.drawImage img 0 0 nil)))
+                                                                                    "png" bos)
+                                                                               (java.io.ByteArrayInputStream. (.toByteArray bos)))))))})))
+
+   Object (fn [x]
+            (if (-> x class .isArray)
+              (seq x)
+              (str x)))})
+
+; datafy and friends shoud be but to good use there
```
Comment (bbatsov) [inline src/clojure/nrepl/rich_printer.clj]: Top-level comments are supposed to have 2+ `;`. At least according to common Lisp lore. :-)

File: src/clojure/nrepl/rich_printer.clj [hunk 8]
```diff
@@ -0,0 +1,460 @@
+(ns nrepl.rich-printer
+  "A printer that produces a pure EDN representation of a Clojure value.
+  Features elisions, MIME attachments and general purpose object browsing.
+  Initially developed for unrepl."
```
Comment (arichiardi) [inline src/clojure/nrepl/rich_printer.clj]: Awesome!

**Comment by cgrand** [comment]
@bbatsov:
> Apart from the build failure 

Do you have any idea why Eastwood does not exit? That's what makes the build to fail.

File: src/clojure/nrepl/elisions.clj (hunk 1 above)
Comment (cgrand) [reply src/clojure/nrepl/elisions.clj]: `elisions` contains the most specific parts: the printer function (for pr-values) and the elision store strategy.

`nrepl.print.elisions` and `nrepl.print.elisions.rich-printer`?

Comment (bbatsov) [reply src/clojure/nrepl/elisions.clj]: Yeah, that sounds good to me.

**Comment by bbatsov** [comment]
> Do you have any idea why Eastwood does not exit? That's what makes the build to fail.

Likely you've hit some bug in it. Does it hang locally as well? Perhaps @slipset might be able to help debugging this?

**Comment by slipset** [comment]
I've reproduced this with the version of eastwood you're running (0.2.6), but cannot reproduce with current eastwood (0.3.4). 0.3.4 has a new linter which catches another error. I'll create a PR with upgraded eastwood and a fix for the error.

**Comment by slipset** [comment]
PR #110 should fix this problem

**Comment by cgrand** [comment]
@slipset The CI still hangs with 0.3.4 https://travis-ci.org/nrepl/nrepl/jobs/477316949

**Comment by bbatsov** [comment]
I'll restart the build. If it hangs again I guess we'll have to disable it until @slipset can help us figuring out what's wrong.

**Comment by cichli** [comment]
Can the eliding phase be separated from the printing phase? So that, for example, other middleware/ops could inspect and use the elided value? Or a custom pretty printer could be used?

**Comment by bbatsov** [comment]
That's a good point!

@cgrand ping :-)

**Comment by cgrand** [comment]
@bbatsov pong!
For elision to be decoupled from printing, printing should be decoupled from traversal.
Which brings us to what @bbatsov nicknamed `printify` recently: a function that given a value returns a representation of this value in a print-safe subset (more or less EDN). (Could `datafy` be used for that?)

There's a second consideration: representing elisions inline or outline. Unrepl represents them inline but it could be conceivable that a more text-based approach would prefer to first have the plain text and after the plain text a list of offsets+length+elisions telling which parts of the original output are elisions ("hypermedia affordances").

**Comment by bbatsov** [comment]
The outline approach seems interesting, but I want to make sure I understand well the idea. @cgrand Can you suggest how a response would look if we opted for the outline approach?

**Comment by cgrand** [comment]
@bbatsov first I most certainly meant offline and not outline :-)
Then here are the options as I see them:

1. use side effects (with dynamic binding around `*out*`) to record where elisions are in the text stream; this precludes building the output in a more functional manner as printing is mandatory. At the end you send the text and a list describing where elisions are. This is the "offline" approach.
2. use a more markupy approach (custom ANSI control sequences?). This is the inline approach.

I personally prefer the second option except that a syntax (well an escape sequence) for the markup is required.

**Comment by bbatsov** [comment]
> @bbatsov first I most certainly meant offline and not outline :-)

🤣 

> I personally prefer the second option except that a syntax (well an escape sequence) for the markup is required.

I'm with you on this one. How complex (implementation-wise) do you estimate it would be go in this direction?

**Comment by cgrand** [comment]
>  How complex (implementation-wise) do you estimate it would be go in this direction?

What are we talking about exactly?
 * transforming a value for print
 * designing a markup/escape mechanism (do we need an escape or are we ok with unlikely accidental collision -- like using a private unicode char or an unused ANSI OSC? it would definitely be simpler)

**Comment by bbatsov** [comment]
> What are we talking about exactly?

I was referring to both points, as they differ from the approach we initially took (this PR). Some solution beats no solution (most of the time), but I'd definitely fond of separation of the ellisions and print steps idea. 

> designing a markup/escape mechanism (do we need an escape or are we ok with unlikely accidental collision -- like using a private unicode char or an unused ANSI OSC? it would definitely be simpler)

I'd go with the simpler solution.

**Comment by bbatsov** [comment]
5 years later, I guess it's time to finally close this PR. :-)

**Pull request closed without merging**